File: src/workshop/beforePost.ts
/**
 * What a game's post files changed in a definition, carried on a copy so the
 * copy is post-processed once (issue #3054).
 *
 * The page reads every definition after the game's `unitdefs_post.lua` and
 * `weapondefs_post.lua` have run, and a copied unit or library weapon starts
 * from that read. The copy goes back into the game as a definition of its own,
 * and the post files run over it again. Balanced Annihilation V15.9.8 scales
 * every weapon's `cratermult` by 0.3 there, so a copy holding the scaled value
 * would be scaled a second time.
 *
 * The unitsync worker says what the post files changed, by definition
 * (`beforepost.rs`), and a copy keeps its part of that. The page goes on
 * showing and editing the values the game ends up with. The compiler writes
 * the game's own values back first (`before_post.rs`), so what the game's
 * post-processing then does to the copy is what it did to the source.
 *
 * Two rules decide which entries a copy keeps, and the compiler applies the
 * same two to later edits:
 *
 * - A path the project changed is the project's. Putting the game's value back
 *   under an edit would undo it.
 * - A weapon slot's `def` and `name` are one field. A post file turns the first
 *   into the second, so a slot pointed elsewhere by `name` must not get the
 *   file's `def` back and fire that instead.
 */
import type { PostChange, UnitDefsResult } from "@/content/bindings";
import type { UnitClones } from "./clones";
import { readPath, sameValue } from "./overrides";
import type { WeaponLibrary } from "./weaponLibrary";

export type { PostChange };

/** A path's steps, compared the way {@link overlaps} compares them. */
function steps(path: string): string[] {
  const out = path.split(".").map((step) => step.toLowerCase());
  if (out.length === 3 && out[0] === "weapons" && out[2] === "def")
    out[2] = "name";
  return out;
}

/** Whether two paths name one field, or one lies inside the other. */
export function overlaps(a: string, b: string): boolean {
  const x = steps(a);
  const y = steps(b);
  const n = Math.min(x.length, y.length);
  for (let i = 0; i < n; i += 1) if (x[i] !== y[i]) return false;
  return true;
}

/** A change with only the entries `keep` answers yes to. */
function filtered(
  change: PostChange,
  keep: (path: string) => boolean,
): PostChange {
  const values = Object.fromEntries(
    Object.entries(change.values ?? {}).filter(([path]) => keep(path)),
  );
  const added = (change.added ?? []).filter(keep);
  return {
    ...(Object.keys(values).length > 0 ? { values } : {}),
    ...(added.length > 0 ? { added } : {}),
  };
}

/** Drop every entry covering a path the project edited. */
export function withoutEdited(
  change: PostChange,
  edited: Iterable<string>,
): PostChange {
  const paths = [...edited];
  return filtered(change, (path) => !paths.some((e) => overlaps(path, e)));
}

/**
 * Drop a unit's own name. A copy gets a name of its own (`clones.ts`), and
 * putting the source's back would undo it.
 */
export function withoutNames(change: PostChange): PostChange {
  return filtered(change, (path) => {
    const lower = path.toLowerCase();
    return lower !== "name" && lower !== "humanname";
  });
}

/**
 * The entries inside one table of a definition, as paths into that table:
 * `weapondefs.armcomlaser` out of a unit's change is its weapon's change.
 */
export function under(change: PostChange, prefix: string): PostChange {
  const start = `${prefix.toLowerCase()}.`;
  const inside = (path: string) => path.toLowerCase().startsWith(start);
  const cut = (path: string) => path.slice(start.length);
  const values = Object.fromEntries(
    Object.entries(change.values ?? {})
      .filter(([path]) => inside(path))
      .map(([path, value]) => [cut(path), value]),
  );
  const added = (change.added ?? []).filter(inside).map(cut);
  return {
    ...(Object.keys(values).length > 0 ? { values } : {}),
    ...(added.length > 0 ? { added } : {}),
  };
}

/**
 * The change for one table inside a definition that the project may have
 * edited, such as a weapon a unit carries being copied into the library:
 * {@link under}, less whatever `edited` covers. An edit that replaces the
 * table itself, or anything around it, leaves nothing of the game's to put
 * back.
 */
export function copiedFrom(
  change: PostChange,
  prefix: string,
  edited: Iterable<string>,
): PostChange {
  const depth = prefix.split(".").length;
  const start = `${prefix.toLowerCase()}.`;
  const inner: string[] = [];
  for (const path of edited) {
    if (!overlaps(path, prefix)) continue;
    if (path.split(".").length <= depth) return {};
    inner.push(path.slice(start.length));
  }
  return withoutEdited(under(change, prefix), inner);
}

/**
 * The part of the source's change that still applies to a copy made before
 * coilbox could tell (issue #3054).
 *
 * A copy saved before this holds the values the page read, and nothing says
 * which of them the game's post files made. Its source still does: an entry is
 * kept wherever the copy holds exactly what the source holds now, which is a
 * post-processed value the copy never changed. Anywhere else, the project
 * edited the copy or the game has changed since, and the copy is left as it is.
 */
export function adoptedChange(
  copy: Record<string, unknown>,
  source: Record<string, unknown>,
  change: PostChange,
): PostChange {
  return filtered(change, (path) =>
    sameValue(readPath(copy, path), readPath(source, path)),
  );
}

/**
 * Give every copy saved before issue #3054 what it can be given, or `null`
 * when there is nothing to give, which is every load after the first.
 *
 * Copied units and library weapons with no record of what the game's post
 * files changed take the part of their source's that {@link adoptedChange}
 * says still applies. Only a copy whose source is still in the game: a copy of
 * a copy, or of a unit the game has since dropped, has nothing to compare
 * against and is left as it was. So is everything when the game's read could
 * not say (`read` absent).
 *
 * `units` and `weaponDefs` are the game's own, not the project's copies.
 */
export function adoptBeforePost(
  clones: UnitClones,
  weapons: WeaponLibrary | undefined,
  read: UnitDefsResult["beforePost"],
  units: Record<string, Record<string, unknown>>,
  weaponDefs: Record<string, Record<string, unknown>>,
): { clones: UnitClones; weapons: WeaponLibrary | undefined } | null {
  if (!read) return null;
  let nextClones = clones;
  for (const clone of Object.values(clones)) {
    if (clone.beforePost || clone.source === undefined) continue;
    const source = units[clone.source];
    if (!source) continue;
    const beforePost = withoutNames(
      adoptedChange(clone.def, source, read.units[clone.source] ?? {}),
    );
    nextClones = { ...nextClones, [clone.key]: { ...clone, beforePost } };
  }
  let nextWeapons = weapons;
  for (const weapon of Object.values(weapons ?? {})) {
    if (weapon.beforePost || !weapon.source) continue;
    const source = weaponDefs[weapon.source];
    if (!source) continue;
    const beforePost = adoptedChange(
      weapon.def,
      source,
      read.weaponDefs[weapon.source] ?? {},
    );
    nextWeapons = { ...nextWeapons, [weapon.key]: { ...weapon, beforePost } };
  }
  return nextClones === clones && nextWeapons === weapons
    ? null
    : { clones: nextClones, weapons: nextWeapons };
}

/** Read a change out of untrusted JSON. `undefined` when it is not one. */
export function parsePostChange(value: unknown): PostChange | undefined {
  if (typeof value !== "object" || value === null || Array.isArray(value))
    return undefined;
  const raw = value as Record<string, unknown>;
  const values =
    typeof raw.values === "object" &&
    raw.values !== null &&
    !Array.isArray(raw.values)
      ? (raw.values as Record<string, unknown>)
      : undefined;
  const added = Array.isArray(raw.added)
    ? raw.added.filter((path): path is string => typeof path === "string")
    : undefined;
  return {
    ...(values && Object.keys(values).length > 0 ? { values } : {}),
    ...(added && added.length > 0 ? { added } : {}),
  };
}

File: crates/tauri-plugin-coilbox-workshop/src/before_post.rs
//! Putting a copy back the way the game's own files had it, before the game
//! post-processes it again (issue #3054).
//!
//! A copied unit or library weapon holds the values the game ended up with,
//! because that is what the page reads and shows. The game's post files run
//! over every definition a mutator adds, so a copy written out with those
//! values is post-processed twice. Its [`PostChange`] says what the post files
//! changed, and [`restore`] writes the game's own values back at those paths
//! just before the copy is compiled. The game then does to the copy exactly
//! what it did to the source.
//!
//! A path the project has changed since is left as the project has it. The
//! value the user set is what they meant, and putting the file's value back
//! under it would undo their edit.

use crate::model::PostChange;
use serde_json::Value;

/// Write the game's own values back into `def`, except where a path in
/// `edited` covers the same field.
///
/// Only into tables `def` still has. A path whose parent the project replaced
/// with something else is skipped, since there is nowhere left for the value
/// to go.
pub(crate) fn restore<'a>(
    def: &mut Value,
    change: &PostChange,
    edited: impl IntoIterator<Item = &'a str> + Clone,
) {
    let untouched = |path: &str| !edited.clone().into_iter().any(|e| overlaps(path, e));
    for (path, value) in &change.values {
        if untouched(path) {
            set(def, path, value.clone());
        }
    }
    for path in &change.added {
        if untouched(path) {
            remove(def, path);
        }
    }
}

/// Whether two paths name the same field or one lies inside the other.
///
/// Compared without regard to case, since the engine reads keys that way. A
/// weapon slot's `def` and its `name` count as one field: a game's post file
/// turns the first into the second, so a slot the user pointed somewhere else
/// by `name` must not get the file's `def` back beside it and fire that
/// instead.
fn overlaps(a: &str, b: &str) -> bool {
    let (a, b) = (steps(a), steps(b));
    a.iter().zip(&b).all(|(x, y)| x == y)
}

fn steps(path: &str) -> Vec<String> {
    let mut out: Vec<String> = path.split('.').map(str::to_lowercase).collect();
    if out.len() == 3 && out[0] == "weapons" && out[2] == "def" {
        out[2] = "name".to_string();
    }
    out
}

/// The table or list `step` names inside `value`, if there is one.
fn child<'a>(value: &'a mut Value, step: &str) -> Option<&'a mut Value> {
    match value {
        Value::Object(map) => map.get_mut(step),
        Value::Array(items) => step.parse::<usize>().ok().and_then(|i| items.get_mut(i)),
        _ => None,
    }
}

/// The container holding the last step of `path`, and that step.
fn parent<'a, 'p>(def: &'a mut Value, path: &'p str) -> Option<(&'a mut Value, &'p str)> {
    let (head, last) = match path.rsplit_once('.') {
        Some((head, last)) => (Some(head), last),
        None => (None, path),
    };
    let mut at = def;
    if let Some(head) = head {
        for step in head.split('.') {
            at = child(at, step)?;
        }
    }
    Some((at, last))
}

fn set(def: &mut Value, path: &str, value: Value) {
    let Some((container, last)) = parent(def, path) else {
        return;
    };
    match container {
        Value::Object(map) => {
            map.insert(last.to_string(), value);
        }
        Value::Array(items) => {
            if let Some(slot) = last.parse::<usize>().ok().and_then(|i| items.get_mut(i)) {
                *slot = value;
            }
        }
        _ => {}
    }
}

/// Take a key the post files added back out. Only from a table: a list entry
/// they added would come back as a list of a different length, which
/// [`PostChange::values`] carries whole instead.
fn remove(def: &mut Value, path: &str) {
    if let Some((Value::Object(map), last)) = parent(def, path) {
        map.remove(last);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    use std::collections::BTreeMap;

    fn change(values: Value, added: &[&str]) -> PostChange {
        PostChange {
            values: serde_json::from_value::<BTreeMap<String, Value>>(values).expect("values"),
            added: added.iter().map(|s| s.to_string()).collect(),
        }
    }

    #[test]
    fn the_game_s_own_values_go_back_and_added_keys_come_out() {
        let mut def = json!({
            "weapons": [{ "name": "armcom_armcomlaser", "onlytargetcategory": "NOTSUB" }],
            "weapondefs": { "armcomlaser": { "cratermult": 0.09, "range": 300 } }
        });
        restore(
            &mut def,
            &change(
                json!({ "weapons.0.def": "ARMCOMLASER" }),
                &["weapons.0.name", "weapondefs.armcomlaser.cratermult"],
            ),
            std::iter::empty(),
        );
        assert_eq!(
            def,
            json!({
                "weapons": [{ "def": "ARMCOMLASER", "onlytargetcategory": "NOTSUB" }],
                "weapondefs": { "armcomlaser": { "range": 300 } }
            })
        );
    }

    #[test]
    fn a_field_the_project_changed_keeps_the_project_s_value() {
        let mut def = json!({ "cratermult": 0.3, "weapons": [{ "name": "a_b" }] });
        restore(
            &mut def,
            &change(
                json!({ "cratermult": 1, "weapons.0.def": "B" }),
                &["weapons.0.name"],
            ),
            ["CraterMult", "weapons.0.name"],
        );
        assert_eq!(
            def,
            json!({ "cratermult": 0.3, "weapons": [{ "name": "a_b" }] })
        );
    }

    #[test]
    fn a_table_the_project_replaced_is_left_alone() {
        let mut def = json!({ "weapondefs": "gone" });
        restore(
            &mut def,
            &change(json!({ "weapondefs.laser.cratermult": 1 }), &[]),
            std::iter::empty(),
        );
        assert_eq!(def, json!({ "weapondefs": "gone" }));
    }
}
